File: disciple/win32/win32_icon.c
/* OCaml-Win32
 * win32_icon.c
 * Copyright (c) 2002 by Harry Chomsky
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the Free
 * Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

#include "win32.h"

/********************************************************************/
/* User Interface Services: Resources: Icons */

CAMLprim value copy_icon(value hicon)
{
    HICON result;

    result = CopyIcon(Handle_val(hicon));
    if (!result)
        raise_last_error();
    return alloc_handle(result);
}

CAMLprim value create_icon_native(value hinstance, value size, value planes, value bits_per_pel, value and_bits, value xor_bits)
{
    HICON result;

    result = CreateIcon(
        Handle_val(hinstance),
        Cx(size), Cy(size),
        (BYTE)Int_val(planes),
        (BYTE)Int_val(bits_per_pel),
        raw_val(and_bits),
        raw_val(xor_bits));
    if (!result)
        raise_last_error();
    return alloc_handle(result);
}

CAMLprim value create_icon_bytecode(value* argv, int argn)
{
    return create_icon_native(argv[0], argv[1], argv[2], argv[3], argv[4], argv[5]);
}

CAMLprim value create_icon_indirect(value iconinfo)
{
    ICONINFO ii;
    HICON result;

    iconinfo_val(iconinfo, &ii);
    result = CreateIconIndirect(&ii);
    if (!result)
        raise_last_error();
    return alloc_handle(result);
}

CAMLprim value destroy_icon(value hicon)
{
    if (!DestroyIcon(Handle_val(hicon)))
        raise_last_error();
    return Val_unit;
}

CAMLprim value draw_icon(value hdc, value point, value hicon)
{
    if (!DrawIcon(Handle_val(hdc), X(point), Y(point), Handle_val(hicon)))
        raise_last_error();
    return Val_unit;
}

CAMLprim value draw_icon_ex_native(value hdc, value point, value hicon, value size, value step, value flicker_free, value options)
{
    BOOL result;

    result = DrawIconEx(
        Handle_val(hdc),
        X(point), Y(point),
        Handle_val(hicon),
        Cx(size), Cy(size),
        Int_val(step),
        Handle_val(flicker_free),
        flags_val(options, draw_icon_options));
    if (!result)
        raise_last_error();
    return Val_unit;
}

CAMLprim value draw_icon_ex_bytecode(value* argv, int argn)
{
    return draw_icon_ex_native(argv[0], argv[1], argv[2], argv[3], argv[4], argv[5], argv[6]);
}

CAMLprim value extract_associated_icon(value hinstance, value filename, value index)
{
    CAMLparam3(hinstance, filename, index);
    char s[_MAX_PATH+1];
    WORD w;
    HICON result;
    CAMLlocal2(val, v);

    strncpy(s, String_val(filename), _MAX_PATH);
    s[_MAX_PATH] = '\0';
    w = (WORD)Int_val(index);
    result = ExtractAssociatedIcon(Handle_val(hinstance), s, &w);
    if (!result)
        raise_last_error();
    val = alloc_tuple(3);
    v = alloc_handle(result);
    Store_field(val, 0, v);
    v = copy_string(s);
    Store_field(val, 1, v);
    Store_field(val, 2, Val_int(w));
    CAMLreturn(val);
}

CAMLprim value get_icon_info(value hicon)
{
    ICONINFO ii;

    if (!GetIconInfo(Handle_val(hicon), &ii))
        raise_last_error();
    return alloc_iconinfo(&ii);
}

CAMLprim value get_icon_info_by_name(value name)
{
    ICONINFO ii;

    if (!GetIconInfo((HICON)resname_val(name), &ii))
        raise_last_error();
    return alloc_iconinfo(&ii);
}

CAMLprim value load_icon(value hinstance, value name)
{
    HICON result;

    result = LoadIcon(Handle_val(hinstance), resname_val(name));
    if (!result)
        raise_last_error();
    return alloc_handle(result);
}

/* Messages */

CAMLprim value deliver_wm_erasebkgnd(value handler, value wnd, value msg)
{
    CAMLparam3(handler, wnd, msg);
    CAMLlocal2(h, result);

    h = alloc_handle((HDC)Wparam(msg));
    result = callback3(handler, wnd, msg, h);
    CAMLreturn(result);
}

CAMLprim value return_wm_erasebkgnd(value result)
{
    return copy_int32(Bool_val(result));
}

CAMLprim value deliver_wm_iconerasebkgnd(value handler, value wnd, value msg)
{
    CAMLparam3(handler, wnd, msg);
    CAMLlocal2(h, result);

    h = alloc_handle((HDC)Wparam(msg));
    result = callback3(handler, wnd, msg, h);
    CAMLreturn(result);
}

File: disciple/win32/win32_icon.ml
(* OCaml-Win32
 * win32_icon.ml
 * Copyright (c) 2002 by Harry Chomsky
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the Free
 * Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *)

open Win32_core

(********************************************************************)
(* User Interface Services: Resources: Icons *)

external copy_icon : icon:hicon -> hicon = "copy_icon"

external create_icon :
    inst:hinstance ->
    size:size ->
    planes:int ->
    bits_per_pel:int ->
    and_bits:raw ->
    xor_bits:raw ->
    hicon = "create_icon_bytecode" "create_icon_native"

(* ... create_icon_from_resource, create_icon_from_resource_ex *)

external create_icon_indirect : info:iconinfo -> hicon = "create_icon_indirect"

external destroy_icon : icon:hicon -> unit = "destroy_icon"

external draw_icon : dc:hdc -> pt:point -> icon:hicon -> unit = "draw_icon"

external draw_icon_ex :
    dc:hdc ->
    pt:point ->
    icon:hicon ->
    size:size ->
    step:int ->
    flicker_free:hbrush ->
    options:draw_icon_option list ->
    unit = "draw_icon_ex_bytecode" "draw_icon_ex_native"

external extract_associated_icon :
    inst:hinstance ->
    filename:string ->
    index:int ->
    (hicon * string * int) = "extract_associated_icon"

(* ... extract_icon, extract_icon_ex *)

external get_icon_info : icon:hicon -> iconinfo = "get_icon_info"

external get_icon_info_by_name : name:resource_name -> iconinfo = "get_icon_info_by_name"

external load_icon : inst:hinstance -> name:resource_name -> hicon = "load_icon"

(* ... lookup_icon_id_from_directory, lookup_icon_id_from_directory_ex *)

(* Messages *)

let wm_erasebkgnd = 0x0014
type wm_erasebkgnd_handler =
    wnd:hwnd ->
    msg:message_contents ->
    dc:hdc ->
    bool message_handler_result
external deliver_wm_erasebkgnd :
    handler:wm_erasebkgnd_handler ->
    wnd:hwnd ->
    msg:message_contents ->
    bool message_handler_result = "deliver_wm_erasebkgnd"
external return_wm_erasebkgnd : result:bool -> message_result = "return_wm_erasebkgnd"
let on_wm_erasebkgnd = on_message ~id:wm_erasebkgnd ~deliver:deliver_wm_erasebkgnd ~return:return_wm_erasebkgnd

let wm_iconerasebkgnd = 0x0027
type wm_iconerasebkgnd_handler =
    wnd:hwnd ->
    msg:message_contents ->
    dc:hdc ->
    unit message_handler_result
external deliver_wm_iconerasebkgnd :
    handler:wm_iconerasebkgnd_handler ->
    wnd:hwnd ->
    msg:message_contents ->
    unit message_handler_result = "deliver_wm_iconerasebkgnd"
let on_wm_iconerasebkgnd = on_message ~id:wm_iconerasebkgnd ~deliver:deliver_wm_iconerasebkgnd ~return:return_simple_message
